### PR TITLE
Kcgourishetti/bedrock agents runnable

### PR DESCRIPTION
**Pull request by kcgourishetti**
## Description
This PR introduces a new Bedrock Agents Runnable that allows using [Bedrock Agents](https://docs.aws.amazon.com/bedrock/latest/userguide/agents-returncontrol.html) with return of control functions as tools. This completes the work presented in #91. 

### Usage with AgentExecutor
```python
import boto3

from langchain.agents import AgentExecutor
from langchain_core.tools import tool

from langchain_aws.agents.base import  BedrockAgentsRunnable

def delete_agent(agent_id: str) -> None:
    bedrock_client = boto3.client("bedrock-agent")
    bedrock_client.delete_agent(agentId=agent_id, skipResourceInUseCheck=True)

@tool("AssetDetail::getAssetValue")
def get_asset_value(asset_holder_id: str) -> str:
    """Get the asset value for an owner id"""
    return f"The total asset value for {asset_holder_id} is 100K"

@tool("AssetDetail::getMortgageRate")
def get_mortgage_rate(asset_holder_id: str, asset_value: str) -> str:
    """Get the mortgage rate based on asset value"""
    return (
        f"The mortgage rate for the asset holder id {asset_holder_id}"
        f"with asset value of {asset_value} is 8.87%"
    )

foundation_model = "anthropic.claude-3-sonnet-20240229-v1:0"
tools = [get_asset_value, get_mortgage_rate]
agent = None
try:
    agent = BedrockAgentsRunnable.create_agent(
        agent_name="mortgage_interest_rate_agent",
        agent_resource_role_arn="<agent-resource-role-arn>",
        foundation_model=foundation_model,
        instruction=(
            "You are an agent who helps with getting the mortgage rate based on "
            "the current asset valuation"
        ),
        tools=tools,
    )
    agent_executor = AgentExecutor(agent=agent, tools=tools)  # type: ignore[arg-type]
    output = agent_executor.invoke(
        {"input": "what is my mortgage rate for id AVC-1234"}
    )

    assert output["output"] == (
        "The mortgage rate for the asset holder id AVC-1234 "
        "with an asset value of 100K is 8.87%."
    )
except Exception as ex:
    raise ex
finally:
    if agent:
        delete_agent(agent.agent_id)
```

### Usage with LangGraph
```python

import boto3

from langgraph.graph import END, START, StateGraph
from langgraph.prebuilt.tool_executor import ToolExecutor

from langchain_aws.agents.base import (
    BedrockAgentAction,
    BedrockAgentFinish,
    BedrockAgentsRunnable,
)

def delete_agent(agent_id: str) -> None:
    bedrock_client = boto3.client("bedrock-agent")
    bedrock_client.delete_agent(agentId=agent_id, skipResourceInUseCheck=True)

@tool
def get_weather(location: str = "") -> str:
    """
    Get the weather of a location

    Args:
        location: location of the place
    """
    if location.lower() == "seattle":
        return f"It is raining in {location}"
    return f"It is hot and humid in {location}"

class AgentState(TypedDict):
    input: str
    output: Union[BedrockAgentAction, BedrockAgentFinish, None]
    intermediate_steps: Annotated[
        list[tuple[BedrockAgentAction, str]], operator.add
    ]

def get_weather_agent_node() -> Tuple[BedrockAgentsRunnable, str]:
    foundation_model = "anthropic.claude-3-sonnet-20240229-v1:0"
    tools = [get_weather]
    try:
        agent_resource_role_arn = _create_agent_role(
            agent_region="us-west-2", 
            foundation_model=foundation_model
        )
        agent = BedrockAgentsRunnable.create_agent(
            agent_name="weather_agent",
            agent_resource_role_arn=agent_resource_role_arn,
            foundation_model=foundation_model,
            instruction=(
                "You are an agent who helps with getting weather for a given "
                "location"
            ),
            tools=tools,
        )

        return agent, agent_resource_role_arn
    except Exception as e:
        raise e

agent_runnable, agent_resource_role_arn = get_weather_agent_node()

def run_agent(data):
    agent_outcome = agent_runnable.invoke(data)
    return {"output": agent_outcome}

tool_executor = ToolExecutor([get_weather])

# Define the function to execute tools
def execute_tools(data):
    # Get the most recent output - this is the key added in the `agent` above
    agent_action = data["output"]
    output = tool_executor.invoke(agent_action[0])
    tuple_output = agent_action[0], output
    return {"intermediate_steps": [tuple_output]}

def should_continue(data):
    output_ = data["output"]

    # If the agent outcome is a list of BedrockAgentActions,
    # then we continue to tool execution
    if (
        isinstance(output_, list)
        and len(output_) > 0
        and isinstance(output_[0], BedrockAgentAction)
    ):
        return "continue"

    # If the agent outcome is an AgentFinish, then we return `exit` string
    # This will be used when setting up the graph to define the flow
    if isinstance(output_, BedrockAgentFinish):
        return "end"

    # Unknown output from the agent, end the graph
    return "end"

try:
    # Define a new graph
    workflow = StateGraph(AgentState)

    # Define the two nodes we will cycle between
    workflow.add_node("agent", run_agent)
    workflow.add_node("action", execute_tools)

    # Set the entrypoint as `agent`
    # This means that this node is the first one called
    workflow.add_edge(START, "agent")

    # We now add a conditional edge
    workflow.add_conditional_edges(
        # First, we define the start node. We use `agent`.
        # This means these are the edges taken after the `agent` node is called.
        "agent",
        # Next, we pass in the function that will determine which node
        # will be called next.
        should_continue,
        # Finally we pass in a mapping.
        # The keys are strings, and the values are other nodes.
        # END is a special node marking that the graph should finish.
        # What will happen is we will call `should_continue`, and then the output
        # of that will be matched against the keys in this mapping.
        # The matched node will then be called.
        {
            # If `tools`, then we call the tool node.
            "continue": "action",
            # Otherwise we finish.
            "end": END,
        },
    )

    # We now add a normal edge from `tools` to `agent`.
    # This means that after `tools` is called, `agent` node is called next.
    workflow.add_edge("action", "agent")

    # Finally, we compile it!
    # This compiles it into a LangChain Runnable,
    # meaning you can use it as you would any other runnable
    app = workflow.compile()

    inputs = {"input": "what is the weather in seattle?"}
    final_state = app.invoke(inputs)

    assert isinstance(final_state.get("output", {}), BedrockAgentFinish)
    assert (
        final_state.get("output").return_values["output"]
        == "It is raining in Seattle"
    )
finally:
    if agent_runnable:
        delete_agent(agent_id=agent_runnable.agent_id)
```

### CODE REVIEW

File: libs/aws/langchain_aws/agents/base.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+            session_state = self._parse_intermediate_steps(
+                input.get("intermediate_steps")  # type: ignore[arg-type]
+            )
+            if session_state is not None:
+                agent_input["sessionState"] = session_state
+        else:
+            agent_input["inputText"] = input.get("input", "")
+
+        output = self.client.invoke_agent(**agent_input)
+        return parse_agent_response(output)
```
Comment (3coins) [inline libs/aws/langchain_aws/agents/base.py:228]: Add callbacks related code and message handling.

**Review by 3coins — COMMENTED** [review]
@kcgourishetti 
Thanks for submitting these changes. Some minor suggestions before we merge.

File: libs/aws/langchain_aws/agents/base.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        function_parameters[arg_name] = {
+            'description': arg_details.get('description', arg_details.get('title', arg_name)),
+            'type': arg_details.get('type', 'string'),
+            'required': not bool(arg_details.get('default', None))
+        }
+    return {
+        'description': tool.description,
+        'name': function_name,
+        'parameters': function_parameters
+    }
```
[the remark this answers is not in the document]
Comment (kcgourishetti) [reply libs/aws/langchain_aws/agents/base.py:153]: the create_action_group API payload for function schema doesn't have any other parameters other than the ones defined below:

```
    ...
    functionSchema={
        'functions': [
            {
                'description': 'string',
                'name': 'string',
                'parameters': {
                    'string': {
                        'description': 'string',
                        'required': True|False,
                        'type': 'string'|'number'|'integer'|'boolean'|'array'
                    }
                }
            },
        ]
    },
    parentActionGroupSignature='AMAZON.UserInput'|'AMAZON.CodeInterpreter'
)
```

File: libs/aws/langchain_aws/agents/base.py (hunk 1 above)
Comment (kcgourishetti) [reply libs/aws/langchain_aws/agents/base.py]: done

File: libs/aws/langchain_aws/agents/base.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+_DEFAULT_ACTION_GROUP_NAME = "DEFAULT_AG_"
+_TEST_AGENT_ALIAS_ID = "TSTALIASID"
+
+
+def get_bedrock_agents_runtime_client():
+    bedrock_config = botocore.client.Config(
+        connect_timeout=120,
+        read_timeout=120,
+        retries={"max_attempts": 3},
+        region_name="us-west-2",
```
[the remark this answers is not in the document]
Comment (kcgourishetti) [reply libs/aws/langchain_aws/agents/base.py]: done

File: libs/aws/langchain_aws/agents/base.py [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+            ) from e
+        except Exception as e:
+            raise ValueError(
+                "Could not load credentials to authenticate with AWS client. "
+                "Please check that credentials in the specified "
+                "profile name are valid."
+            ) from e
+
+    @staticmethod
+    def __get_boto_session(credentials_profile_name: str, region_name: str, endpoint_url: str):
```
Comment (3coins) [inline libs/aws/langchain_aws/agents/base.py:311]: Can we move this out as a separate method.

File: libs/aws/tests/integration_tests/agents/bedrock_agent_langgraph.ipynb [hunk 5]
```diff
@@ -0,0 +1,147 @@
+{
```
Comment (3coins) [inline libs/aws/tests/integration_tests/agents/bedrock_agent_langgraph.ipynb:1]: Can you move them to the samples directory.

File: libs/aws/tests/integration_tests/agents/bedrock_agents_roc.ipynb [hunk 6]
```diff
@@ -0,0 +1,383 @@
+{
```
Comment (3coins) [inline libs/aws/tests/integration_tests/agents/bedrock_agents_roc.ipynb:1]: Move them to samples directory.

File: libs/aws/tests/integration_tests/agents/test_bedrock_agents.py [hunk 7]
```diff
@@ -0,0 +1,411 @@
+import json
```
Comment (3coins) [inline libs/aws/tests/integration_tests/agents/test_bedrock_agents.py:1]: We don't have the permissions to set up the agents and action groups. Can we add the @skip decorator for the tests now.

**Review by 3coins — COMMENTED** [review]
@Kcgourishetti
Great job on all the changes. We are almost there, some minor comments on aligning the `create_job` with the [create_agent](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/bedrock-agent/client/create_agent.html) API, and moving notebooks to samples directory.

File: libs/aws/langchain_aws/agents/base.py [hunk 8]
```diff
@@ -0,0 +1,514 @@
...
+            instructions: str,
+            tools: List[BaseTool] = [],
+            *,
+            memory_storage_days: int = 0,
+            guardrail_id: str = None,
+            guardrail_version: str = "DRAFT",
+            credentials_profile_name: str = None,
+            region_name: str = None,
+            bedrock_endpoint_url: str = None,
+            runtime_endpoint_url: str = None,
```
[the remark this answers is not in the document]
Comment (kcgourishetti) [reply libs/aws/langchain_aws/agents/base.py:346]: yes, one is the endpoint for control plane responsible for CRUDL of the agents and one is the runtime responsible for invoke etc. the boto client names for each of these are also different one is `bedrock-agent` and other is `bedrock-agent-runtime`.

File: libs/aws/langchain_aws/agents/base.py [hunk 9]
```diff
@@ -0,0 +1,514 @@
...
+        return client_params, session
+
+    @classmethod
+    def create_agent(
+            cls,
+            agent_name: str,
+            agent_resource_role_arn: str,
+            model: str,
+            instructions: str,
+            tools: List[BaseTool] = [],
```
[the remark this answers is not in the document]
Comment (kcgourishetti) [reply libs/aws/langchain_aws/agents/base.py:338]: bedrock agent doesn't need tools. when tools are not provided, it works like a Q&A agent. same is true with LangChain agent. you can pass in empty list of tools to the agent executor and it works like a Q&A agent.

File: libs/aws/langchain_aws/agents/base.py [hunk 10]
```diff
@@ -0,0 +1,514 @@
...
+            client_params["endpoint_url"] = endpoint_url
+        return client_params, session
+
+    @classmethod
+    def create_agent(
+            cls,
+            agent_name: str,
+            agent_resource_role_arn: str,
+            model: str,
+            instructions: str,
```
[the remark this answers is not in the document]
Comment (kcgourishetti) [reply libs/aws/langchain_aws/agents/base.py:337]: you are right. except agent_name nothing else is required. but that kind of agent is literally unusable. for the invoke to work, it needs a model, agent role, AND instructions..without which neither prepare will work nor invoke. following the api conventions here may not be good here. let me know your thoughts.

File: libs/aws/langchain_aws/agents/base.py [hunk 11]
```diff
@@ -0,0 +1,514 @@
...
+            tools: List[BaseTool] = [],
+            *,
+            memory_storage_days: int = 0,
```
[the remark this answers is not in the document]
Comment (kcgourishetti) [reply libs/aws/langchain_aws/agents/base.py]: as per our conversation, removed this for now. we will add it once the agent memory delay is within normal bounds.

File: libs/aws/langchain_aws/agents/base.py (hunk 4 above)
Comment (kcgourishetti) [reply libs/aws/langchain_aws/agents/base.py]: done

File: libs/aws/tests/integration_tests/agents/bedrock_agent_langgraph.ipynb (hunk 5 above)
Comment (kcgourishetti) [reply libs/aws/tests/integration_tests/agents/bedrock_agent_langgraph.ipynb]: done

File: libs/aws/tests/integration_tests/agents/bedrock_agents_roc.ipynb (hunk 6 above)
Comment (kcgourishetti) [reply libs/aws/tests/integration_tests/agents/bedrock_agents_roc.ipynb]: done

File: libs/aws/tests/integration_tests/agents/test_bedrock_agents.py (hunk 7 above)
Comment (kcgourishetti) [reply libs/aws/tests/integration_tests/agents/test_bedrock_agents.py:1]: done.

File: libs/aws/langchain_aws/agents/base.py [hunk 12]
```diff
@@ -0,0 +1,514 @@
...
+            guardrail_id: str = None,
+            guardrail_version: str = "DRAFT",
```
[the remark this answers is not in the document]
Comment (kcgourishetti) [reply libs/aws/langchain_aws/agents/base.py]: done.

File: libs/aws/langchain_aws/agents/base.py [hunk 13]
```diff
@@ -0,0 +1,514 @@
...
+            credentials_profile_name: str = None,
+            region_name: str = None,
+            bedrock_endpoint_url: str = None,
+            runtime_endpoint_url: str = None,
+            **kwargs: Any,
+    ) -> BedrockAgentsRunnable:
+        """
```
[the remark this answers is not in the document]
Comment (kcgourishetti) [reply libs/aws/langchain_aws/agents/base.py]: took at shot at it. please take a look. happy to address if you have any more comments on it.

**Comment by 3coins** [comment]
@kcgourishetti 
I took care of some of the formatting, linting, typing and CI errors. One last thing I noticed is that the notebook `bedrock_agent_langgraph.ipynb` seems to be incomplete, it doesn't have any of the code for the LangGraph executor, you seem to have some of this code in the integration test.

A couple of other observations, not blockers on this PR:
1. Even though there is no LLM present here (agent manages the LLM calls), a lot of the LangChain components work around `messages` which is central to it's working. We should follow up with a PR to support messages.
2. The LangGraph integration test is using the [ToolExecutor](https://github.com/langchain-ai/langgraph/blob/main/libs/langgraph/langgraph/prebuilt/tool_executor.py#L17), which has been deprecated in favor of [ToolNode](https://github.com/langchain-ai/langgraph/blob/main/libs/langgraph/langgraph/prebuilt/tool_node.py#L42), which works only with messages. To future proof this work, we should move to use ToolNode in LangGraph samples and tests.

**Comment by kcgourishetti** [comment]
> @kcgourishetti I took care of some of the formatting, linting, typing and CI errors. One last thing I noticed is that the notebook `bedrock_agent_langgraph.ipynb` seems to be incomplete, it doesn't have any of the code for the LangGraph executor, you seem to have some of this code in the integration test.
> 
> A couple of other observations, not blockers on this PR:
> 
> 1. Even though there is no LLM present here (agent manages the LLM calls), a lot of the LangChain components work around `messages` which is central to it's working. We should follow up with a PR to support messages.
> 2. The LangGraph integration test is using the [ToolExecutor](https://github.com/langchain-ai/langgraph/blob/main/libs/langgraph/langgraph/prebuilt/tool_executor.py#L17), which has been deprecated in favor of [ToolNode](https://github.com/langchain-ai/langgraph/blob/main/libs/langgraph/langgraph/prebuilt/tool_node.py#L42), which works only with messages. To future proof this work, we should move to use ToolNode in LangGraph samples and tests.

- `bedrock_agent_langgraph.ipynb`: weird. looks like the file got overwritten. updated and committed it now.
- `messages`: ack. happy to hear if you have any thoughts on it.
- `ToolExecutor`: ah, thats a nice catch. we gotta make this work with messages then. thanks for the detailed info.

**Pull request merged**